exposure-notification/src/main/java/fi/thl/covid19/exposurenotification/diagnosiskey/DiagnosisKeyDao.java

```diff
@@ -109,39 +112,53 @@ public int getKeyCountV2(int intervalV2) {
     @Transactional
     public List<TemporaryExposureKey> getIntervalKeys(int interval) {
         LOG.info("Fetching keys: {}", keyValue("interval", interval));
-        String sql =
-                "select key_data, rolling_period, rolling_start_interval_number, transmission_risk_level, " +
-                        "submission_interval, submission_interval_v2, " +
-                        "origin, visited_countries, days_since_onset_of_symptoms, consent_to_share, symptoms_exist " +
-                        "from en.diagnosis_key " +
-                        "where submission_interval = :interval " +
-                        // Level 0 & 7 would get 0 score anyhow, so ignore them
-                        // This also clips the range, so that we can manage the difference between iOS & Android APIs
-                        "and transmission_risk_level between 1 and 6 " +
-                        "order by key_data";
+        String sql = "select key_data, rolling_period, rolling_start_interval_number, transmission_risk_level, " +
+                "submission_interval, submission_interval_v2, " +
+                "origin, visited_countries, days_since_onset_of_symptoms, consent_to_share, symptoms_exist " +
+                "from en.diagnosis_key " +
+                "where submission_interval = :interval " +
+                // Level 0 & 7 would get 0 score anyhow, so ignore them
+                // This also clips the range, so that we can manage the difference between iOS & Android APIs
+                "and transmission_risk_level between 1 and 6 " +
+                "order by key_data";
         Map<String, Object> params = Map.of("interval", interval);
         // We should not have invalid data in the DB, but if we do, pass by it and move on
-        return jdbcTemplate.query(sql, params, (rs, i) -> mapValidKey(interval, rs, i))
+        List<TemporaryExposureKey> keys = jdbcTemplate.query(sql, params, (rs, i) -> mapValidKey(interval, rs, i))
                 .stream().flatMap(Optional::stream).collect(Collectors.toList());
+
+        if (keys.isEmpty() || keys.size() >= BATCH_MIN_SIZE) {
+            return keys;
+        } else {
+            List<TemporaryExposureKey> dummyKeys = generateDummyKeys(BATCH_MIN_SIZE - keys.size(), false, from24hourToV2Interval(interval), Instant.now());
+            batchInsert(dummyKeys, Optional.empty());
```

Comment:
I dislike that a getSomething method has a side effect of inserting stuff into the database. Perhaps this could be a separate function, for example padKeysForInterval(int interval) that just does the count-check and inserts dummies. That can then be called before the actual fetch for the same effect but keeping the concerns separate (and separately testable).

Reply:
That would also avoid having the near-identical code in all of these fetches

Reply:
Yep, good points and good discussions on other channels. I pushed kind of compromise solution just a moment ago.